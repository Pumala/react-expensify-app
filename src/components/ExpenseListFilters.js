import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused });
    };
    onDatesChange = ({ startDate, endDate}) => {
        
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);        
    };
    onSortByChange = (e) => {
        const value = e.target.value;

        if (value == 'date') {
            this.props.sortByDate();
        } else {
            this.props.sortByAmount();
        }
    };
    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextFilterChange} />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}>
                    <option>Select</option>        
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>            
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={false}
                    numberOfMonths={1}
                    isOutsideRange={() => false }
                />
            </div>
        );
    };
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);