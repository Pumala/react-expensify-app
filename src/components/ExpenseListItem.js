import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
    <li>
        <p>ID: {id}</p>
        <Link to={`/edit/${id}`}>
            <p>Description: {description}</p>
        </Link>
        <p>Note: {note}</p>
        <p>
            Amount: {numeral(amount / 100).format('$0,0.00')}
            <br />
            Created At: {moment(createdAt).format("MMMM Do, YYYY ")}
        </p>
        <button>Expense Item Page</button>
    </li>
); 

export default ExpenseListItem;