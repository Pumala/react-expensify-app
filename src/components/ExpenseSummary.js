import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount > 1 ? 'expenses' : 'expense';
    const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div>
            <p>Expenses Summary El</p>
            {expensesCount > 0 && <p>
                You are viewing {expensesCount} {expenseWord} totaling {formattedExpenseTotal}.
            </p>}
            {expensesCount < 1 && <p>There are no expenses at this time.</p>}
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);