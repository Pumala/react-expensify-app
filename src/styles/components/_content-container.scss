.content-container {
    max-width: 80rem;
    margin: auto;
    padding: 0 $m-size;
    align-items: center;
}