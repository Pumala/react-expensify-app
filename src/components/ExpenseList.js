import React from 'react';
// called connect because it connects your component to the redux store
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    return (
        <div>
            <h1>Expense List</h1>
            <ul>
            {
                props.expenses.length === 0 ? (
                    <li>No expenses found.</li>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
        
                        // expense={expense} or {...expense}
                        // similar, not the same
                    })
                )
            }
            </ul>
        </div>
    )
};

// this function lets us determine what info we want from the store that 
    // we want our component to access
const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

// export HOC
export default connect(mapStateToProps)(ExpenseList); 



// Alt below
// HOC
// const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

// export default ConnectedExpenseList;