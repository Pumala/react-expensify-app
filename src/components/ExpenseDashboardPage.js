import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => (
    <div>
        <h1>This is from Dashboard Component.</h1>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;