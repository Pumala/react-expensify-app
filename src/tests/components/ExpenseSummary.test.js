import React from 'react';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import { shallow } from 'enzyme';

test('should render Expense Summary correctly with no expense', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={0} expensesTotal={0} />);

    expect(wrapper).toMatchSnapshot();
});

test('should render Expense Summary correctly with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={1} expensesTotal={235} />);

    expect(wrapper).toMatchSnapshot();
    
});

test('should render Expense Summary correctly with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={23} expensesTotal={4758374} />);

    expect(wrapper).toMatchSnapshot();
    
});