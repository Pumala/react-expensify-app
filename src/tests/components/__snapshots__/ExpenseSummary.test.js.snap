// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense Summary correctly with 1 expense 1`] = `
<div>
  <p>
    Expenses Summary El
  </p>
  <p>
    You are viewing 
    1
     
    expense
     totaling 
    $2.35
    .
  </p>
</div>
`;

exports[`should render Expense Summary correctly with multiple expenses 1`] = `
<div>
  <p>
    Expenses Summary El
  </p>
  <p>
    You are viewing 
    23
     
    expenses
     totaling 
    $47,583.74
    .
  </p>
</div>
`;

exports[`should render Expense Summary correctly with no expense 1`] = `
<div>
  <p>
    Expenses Summary El
  </p>
  <p>
    There are no expenses at this time.
  </p>
</div>
`;
